compose/project.py

```diff
@@ -695,7 +695,9 @@ def push(self, service_names=None, ignore_push_failures=False):
             repo, tag, sep = parse_repository_tag(service.image_name)
             service_image_name = sep.join((repo, tag)) if tag else sep.join((repo, 'latest'))
 
-            if service_image_name not in unique_images:
+            # Add service to unique if it's pushable
+            if (service_image_name not in unique_images and
```

Comment:
With my suggestion to modify `service.image_name` above, this function could look something like;

```python
def push(self, service_names=None, ignore_push_failures=False):
    unique_images = set()
    for service in self.get_services(service_names, include_deps=False):
        if service.can_be_pushed() and service.image_name not in unique_images:
            service.push(ignore_push_failures)
            unique_images.add(service.image_name)
```

Reply:
Actually, given that `service.push()` itself checks `can_be_pushed`, we could just skip that check here, and delegate it to `service.push()`;

```python
def push(self, service_names=None, ignore_push_failures=False):
    unique_images = set()
    for service in self.get_services(service_names, include_deps=False):
        if service.image_name not in unique_images:
            service.push(ignore_push_failures)
            unique_images.add(service.image_name)
```

Reply:
> Actually, given that `service.push()` itself checks `can_be_pushed`, we could just skip that check here, and delegate it to `service.push()`;
That's what caused the original issue :)
Services that were not supposed to be pushed tried to push themselves, and added their `image` to `unique_images`

Reply:
But, we can make `service.push` return None if image can't be pushed. And rely on it:
```python
def push(self, service_names=None, ignore_push_failures=False):
    unique_images = set()
    for service in self.get_services(service_names, include_deps=False):
        if service.push(ignore_push_failures) and service.image_name not in unique_images:
            unique_images.add(service.image_name)
```

Reply:
> Services that were not supposed to be pushed tried to push themselves, and added their image to unique_images

🤦 you're right, my mistake


> But, we can make service.push return None if it can't be pushed. (of course, document it and add test). And rely on it like this:

not sure if that's clearer; I think we should take the original approach (`if service.can_be_pushed() and service.image_name not in unique_images`) I think that properly describes the intent